File: packages/less/lib/less/tree/atrule.js
// @ts-check
/** @import { EvalContext, CSSOutput, TreeVisitor, FileInfo, VisibilityInfo } from './node.js' */
/** @import { FunctionRegistry } from './nested-at-rule.js' */
import Node from './node.js';
import Selector from './selector.js';
import Ruleset from './ruleset.js';
import Anonymous from './anonymous.js';
import NestableAtRulePrototype from './nested-at-rule.js';
import mergeRules from './merge-rules.js';

/**
 * @typedef {Node & {
 *   rules?: Node[],
 *   selectors?: Selector[],
 *   root?: boolean,
 *   allowImports?: boolean,
 *   functionRegistry?: FunctionRegistry,
 *   merge?: boolean,
 *   debugInfo?: { lineNumber: number, fileName: string },
 *   elements?: import('./element.js').default[]
 * }} RulesetLikeNode
 */

class AtRule extends Node {
    get type() { return 'AtRule'; }

    /**
     * @param {string} [name]
     * @param {Node | string} [value]
     * @param {Node[] | Ruleset} [rules]
     * @param {number} [index]
     * @param {FileInfo} [currentFileInfo]
     * @param {{ lineNumber: number, fileName: string }} [debugInfo]
     * @param {boolean} [isRooted]
     * @param {VisibilityInfo} [visibilityInfo]
     */
    constructor(
        name,
        value,
        rules,
        index,
        currentFileInfo,
        debugInfo,
        isRooted,
        visibilityInfo
    ) {
        super();
        let i;
        var selectors = (new Selector([], null, null, index, currentFileInfo)).createEmptySelectors();

        /** @type {string | undefined} */
        this.name  = name;
        this.value = (value instanceof Node) ? value : (value ? new Anonymous(value) : value);
        /** @type {boolean | undefined} */
        this.simpleBlock = undefined;
        /** @type {RulesetLikeNode[] | undefined} */
        this.declarations = undefined;
        /** @type {RulesetLikeNode[] | undefined} */
        this.rules = undefined;
        if (rules) {
            if (Array.isArray(rules)) {
                const allDeclarations = this.declarationsBlock(rules);

                let allRulesetDeclarations = true;
                rules.forEach(rule => {
                    if (rule.type === 'Ruleset' && /** @type {RulesetLikeNode} */ (rule).rules) allRulesetDeclarations = allRulesetDeclarations && this.declarationsBlock(/** @type {Node[]} */ (/** @type {RulesetLikeNode} */ (rule).rules), true);
                });

                if (allDeclarations && !isRooted) {
                    this.simpleBlock = true;
                    this.declarations = rules;
                } else if (allRulesetDeclarations && rules.length === 1 && !isRooted && !value) {
                    this.simpleBlock = true;
                    this.declarations = /** @type {RulesetLikeNode} */ (rules[0]).rules ? /** @type {RulesetLikeNode} */ (rules[0]).rules : rules;
                } else {
                    this.rules = rules;
                }
            } else {
                const allDeclarations = this.declarationsBlock(/** @type {Node[]} */ (rules.rules));

                if (allDeclarations && !isRooted && !value) {
                    this.simpleBlock = true;
                    this.declarations = rules.rules;
                } else {
                    this.rules = [rules];
                    /** @type {RulesetLikeNode} */ (this.rules[0]).selectors = (new Selector([], null, null, index, currentFileInfo)).createEmptySelectors();
                }
            }
            if (!this.simpleBlock) {
                for (i = 0; i < this.rules.length; i++) {
                    /** @type {RulesetLikeNode} */ (this.rules[i]).allowImports = true;
                }
            }
            this.setParent(selectors, /** @type {Node} */ (/** @type {unknown} */ (this)));
            this.setParent(this.rules, /** @type {Node} */ (/** @type {unknown} */ (this)));
        }
        this._index = index;
        this._fileInfo = currentFileInfo;
        /** @type {{ lineNumber: number, fileName: string } | undefined} */
        this.debugInfo = debugInfo;
        /** @type {boolean} */
        this.isRooted = isRooted || false;
        this.copyVisibilityInfo(visibilityInfo);
        this.allowRoot = true;
    }

    /**
     * @param {Node[]} rules
     * @param {boolean} [mergeable]
     * @returns {boolean}
     */
    declarationsBlock(rules, mergeable = false) {
        if (!mergeable) {
            return rules.filter(function (/** @type {Node & { merge?: boolean }} */ node) { return (node.type === 'Declaration' || node.type === 'Comment') && !node.merge}).length === rules.length;
        } else {
            return rules.filter(function (/** @type {Node} */ node) { return (node.type === 'Declaration' || node.type === 'Comment'); }).length === rules.length;
        }
    }

    /**
     * @param {Node[]} rules
     * @returns {boolean}
     */
    keywordList(rules) {
        if (!Array.isArray(rules)) {
            return false;
        } else {
            return rules.filter(function (/** @type {Node} */ node) { return (node.type === 'Keyword' || node.type === 'Comment'); }).length === rules.length;
        }
    }

    /** @param {TreeVisitor} visitor */
    accept(visitor) {
        const value = this.value, rules = this.rules, declarations = this.declarations;

        if (rules) {
            this.rules = visitor.visitArray(rules);
        } else if (declarations) {
            this.declarations = visitor.visitArray(declarations);
        }
        if (value) {
            this.value = visitor.visit(/** @type {Node} */ (value));
        }
    }

    /** @override @returns {boolean} */
    isRulesetLike() {
        return /** @type {boolean} */ (/** @type {unknown} */ (this.rules || !this.isCharset()));
    }

    isCharset() {
        return '@charset' === this.name;
    }

    /**
     * @param {EvalContext} context
     * @param {CSSOutput} output
     */
    genCSS(context, output) {
        const value = this.value, rules = this.rules || this.declarations;
        output.add(/** @type {string} */ (this.name), this.fileInfo(), this.getIndex());
        if (value) {
            output.add(' ');
            /** @type {Node} */ (value).genCSS(context, output);
        }
        if (this.simpleBlock) {
            this.outputRuleset(context, output, /** @type {Node[]} */ (this.declarations));
        } else if (rules) {
            this.outputRuleset(context, output, rules);
        } else {
            output.add(';');
        }
    }

    /**
     * @param {EvalContext} context
     * @returns {Node}
     */
    eval(context) {
        let mediaPathBackup, mediaBlocksBackup, value = this.value, rules = this.rules || this.declarations;

        // media stored inside other atrule should not bubble over it
        // backpup media bubbling information
        mediaPathBackup = context.mediaPath;
        mediaBlocksBackup = context.mediaBlocks;
        // deleted media bubbling information
        context.mediaPath = [];
        context.mediaBlocks = [];

        if (value) {
            value = /** @type {Node} */ (value).eval(context);
        }

        if (rules) {
            rules = this.evalRoot(context, rules);
        }
        if (Array.isArray(rules) && /** @type {RulesetLikeNode} */ (rules[0]).rules && Array.isArray(/** @type {RulesetLikeNode} */ (rules[0]).rules) && /** @type {Node[]} */ (/** @type {RulesetLikeNode} */ (rules[0]).rules).length) {
            const allMergeableDeclarations = this.declarationsBlock(/** @type {Node[]} */ (/** @type {RulesetLikeNode} */ (rules[0]).rules), true);
            if (allMergeableDeclarations && !this.isRooted && !value) {
                mergeRules(/** @type {Node[]} */ (/** @type {RulesetLikeNode} */ (rules[0]).rules));
                rules = /** @type {RulesetLikeNode[]} */ (/** @type {RulesetLikeNode} */ (rules[0]).rules);
                rules.forEach(/** @param {RulesetLikeNode} rule */ rule => { rule.merge = false; });
            }
        }
        if (this.simpleBlock && rules) {
            /** @type {RulesetLikeNode} */ (rules[0]).functionRegistry = /** @type {RulesetLikeNode} */ (context.frames[0]).functionRegistry.inherit();
            rules = rules.map(function (/** @type {Node} */ rule) { return rule.eval(context); });
        }

        // restore media bubbling information
        context.mediaPath = mediaPathBackup;
        context.mediaBlocks = mediaBlocksBackup;
        return /** @type {Node} */ (/** @type {unknown} */ (new AtRule(this.name, value, rules, this.getIndex(), this.fileInfo(), this.debugInfo, this.isRooted, this.visibilityInfo())));
    }

    /**
     * @param {EvalContext} context
     * @param {Node[]} rules
     * @returns {Node[]}
     */
    evalRoot(context, rules) {
        let ampersandCount = 0;
        let noAmpersandCount = 0;
        let noAmpersands = true;

        if (!this.simpleBlock) {
            rules = [rules[0].eval(context)];
        }

        /** @type {Selector[]} */
        let precedingSelectors = [];
        if (context.frames.length > 0) {
            for (let index = 0; index < context.frames.length; index++) {
                const frame = /** @type {RulesetLikeNode} */ (context.frames[index]);
                if (
                    frame.type === 'Ruleset' &&
                    frame.rules &&
                    frame.rules.length > 0
                ) {
                    if (frame && !frame.root && frame.selectors && frame.selectors.length > 0) {
                        precedingSelectors = precedingSelectors.concat(frame.selectors);
                    }
                }
                if (precedingSelectors.length > 0) {
                    const allAmpersandElements = precedingSelectors.every(
                        sel => sel.elements && sel.elements.length > 0 && sel.elements.every(
                            /** @param {import('./element.js').default} el */
                            el => el.value === '&'
                        )
                    );
                    if (allAmpersandElements) {
                        noAmpersands = false;
                        noAmpersandCount++;
                    } else {
                        ampersandCount++;
                    }
                }
            }
        }

        const mixedAmpersands = ampersandCount > 0 && noAmpersandCount > 0 && !noAmpersands;
        if (
            (this.isRooted && ampersandCount > 0 && noAmpersandCount === 0 && noAmpersands)
            || !mixedAmpersands
        ) {
            /** @type {RulesetLikeNode} */ (rules[0]).root = true;
        }
        return rules;
    }

    /** @param {string} name */
    variable(name) {
        if (this.rules) {
            // assuming that there is only one rule at this point - that is how parser constructs the rule
            return Ruleset.prototype.variable.call(this.rules[0], name);
        }
    }

    find() {
        if (this.rules) {
            // assuming that there is only one rule at this point - that is how parser constructs the rule
            return Ruleset.prototype.find.apply(this.rules[0], arguments);
        }
    }

    rulesets() {
        if (this.rules) {
            // assuming that there is only one rule at this point - that is how parser constructs the rule
            return Ruleset.prototype.rulesets.apply(this.rules[0]);
        }
    }

    /**
     * @param {EvalContext} context
     * @param {CSSOutput} output
     * @param {Node[]} rules
     */
    outputRuleset(context, output, rules) {
        const ruleCnt = rules.length;
        let i;
        context.tabLevel = (context.tabLevel | 0) + 1;

        // Compressed
        if (context.compress) {
            output.add('{');
            for (i = 0; i < ruleCnt; i++) {
                rules[i].genCSS(context, output);
            }
            output.add('}');
            context.tabLevel--;
            return;
        }

        // Non-compressed
        const tabSetStr = `\n${Array(context.tabLevel).join('  ')}`, tabRuleStr = `${tabSetStr}  `;
        if (!ruleCnt) {
            output.add(` {${tabSetStr}}`);
        } else {
            output.add(` {${tabRuleStr}`);
            rules[0].genCSS(context, output);
            for (i = 1; i < ruleCnt; i++) {
                output.add(tabRuleStr);
                rules[i].genCSS(context, output);
            }
            output.add(`${tabSetStr}}`);
        }

        context.tabLevel--;
    }
}

// Apply shared methods from NestableAtRulePrototype that AtRule doesn't override
const { evalFunction, evalTop, evalNested, permute, bubbleSelectors } = NestableAtRulePrototype;
Object.assign(AtRule.prototype, { evalFunction, evalTop, evalNested, permute, bubbleSelectors });

export default AtRule;

File: packages/less/src/less/less_go/atrule.go
package less_go

import (
	"fmt"
	"strings"
)

// AtRule represents an at-rule node in the Less AST
type AtRule struct {
	*Node
	Name        string
	Value       any
	Rules       []any
	IsRooted    bool
	AllowRoot   bool
	DebugInfo   any
	AllExtends  []*Extend // For storing extends found by ExtendFinderVisitor
}

// NewAtRule creates a new AtRule instance
func NewAtRule(name string, value any, rules any, index int, currentFileInfo map[string]any, debugInfo any, isRooted bool, visibilityInfo map[string]any) *AtRule {
	node := NewNode()
	node.TypeIndex = GetTypeIndexForNodeType("AtRule")

	atRule := &AtRule{
		Node:      node,
		Name:      name,
		IsRooted:  isRooted,
		AllowRoot: true,
		DebugInfo: debugInfo,
	}

	// Handle value - convert to Anonymous if string/non-Node
	if value != nil {
		// Check if value is already a Node-based type
		// This matches JavaScript: (value instanceof Node) ? value : (value ? new Anonymous(value) : value)
		// Preserve Anonymous nodes and any node that has a GetType() method (Quoted, Variable, Expression, etc.)
		if _, ok := value.(*Anonymous); ok {
			atRule.Value = value
		} else if _, ok := value.(interface{ GetType() string }); ok {
			atRule.Value = value
		} else {
			atRule.Value = NewAnonymous(value, index, currentFileInfo, false, false, nil)
		}
	} else {
		atRule.Value = value
	}

	// Handle rules
	if rules != nil {
		if rulesSlice, ok := rules.([]any); ok {
			atRule.Rules = rulesSlice
		} else {
			// Single rule - convert to array and add empty selectors
			atRule.Rules = []any{rules}
			if ruleset, ok := rules.(*Ruleset); ok {
				// Create empty selectors like JavaScript version - skip if errors occur
				emptySelector, err := NewSelector("", nil, nil, index, currentFileInfo, nil)
				if err == nil {
					emptySelectors, err := emptySelector.CreateEmptySelectors()
					if err == nil {
						ruleset.Selectors = make([]any, len(emptySelectors))
						for i, sel := range emptySelectors {
							ruleset.Selectors[i] = sel
						}
					}
				}
			}
		}

		// Set allowImports to true for all rules and parent relationships
		for _, rule := range atRule.Rules {
			if ruleset, ok := rule.(*Ruleset); ok {
				ruleset.AllowImports = true
			}
		}
		atRule.SetParent(atRule.Rules, atRule.Node)
	}

	// Set node properties
	atRule.Index = index
	if currentFileInfo != nil {
		atRule.SetFileInfo(currentFileInfo)
	}
	atRule.CopyVisibilityInfo(visibilityInfo)

	return atRule
}

// Type returns the node type
func (a *AtRule) Type() string {
	return "AtRule"
}

// GetType returns the node type
func (a *AtRule) GetType() string {
	return "AtRule"
}

// GetName returns the at-rule name
func (a *AtRule) GetName() string {
	return a.Name
}

// GetDebugInfo returns debug info for the at-rule
func (a *AtRule) GetDebugInfo() any {
	return a.DebugInfo
}

// ToCSS converts the at-rule to CSS string
func (a *AtRule) ToCSS(context any) string {
	var strs []string
	output := &CSSOutput{
		Add: func(chunk any, fileInfo any, index any) {
			strs = append(strs, fmt.Sprintf("%v", chunk))
		},
		IsEmpty: func() bool {
			return len(strs) == 0
		},
	}
	a.GenCSS(context, output)
	return strings.Join(strs, "")
}

// Accept visits the node with a visitor
func (a *AtRule) Accept(visitor any) {
	if v, ok := visitor.(interface{ VisitArray([]any) []any }); ok {
		if a.Rules != nil {
			a.Rules = v.VisitArray(a.Rules)
		}
	}

	if v, ok := visitor.(interface{ Visit(any) any }); ok {
		if a.Value != nil {
			a.Value = v.Visit(a.Value)
		}
	}
}

// IsRulesetLike checks if the at-rule is ruleset-like
func (a *AtRule) IsRulesetLike() any {
	if a.Rules != nil {
		return a.Rules
	}
	return !a.IsCharset()
}

// IsCharset checks if this is a @charset rule
func (a *AtRule) IsCharset() bool {
	return a.Name == "@charset"
}

// GenCSS generates CSS representation
func (a *AtRule) GenCSS(context any, output *CSSOutput) {
	output.Add(a.Name, a.FileInfo(), a.GetIndex())
	
	if a.Value != nil {
		output.Add(" ", nil, nil)
		if gen, ok := a.Value.(interface{ GenCSS(any, *CSSOutput) }); ok {
			gen.GenCSS(context, output)
		}
	}
	
	if a.Rules != nil {
		a.OutputRuleset(context, output, a.Rules)
	} else {
		output.Add(";", nil, nil)
	}
}

// Eval evaluates the at-rule
func (a *AtRule) Eval(context any) (any, error) {
	var mediaPathBackup, mediaBlocksBackup any
	var value any = a.Value
	var rules []any = a.Rules

	// Media stored inside other atrule should not bubble over it
	// Backup media bubbling information
	if ctx, ok := context.(map[string]any); ok {
		mediaPathBackup = ctx["mediaPath"]
		mediaBlocksBackup = ctx["mediaBlocks"]
		// Delete media bubbling information
		ctx["mediaPath"] = []any{}
		ctx["mediaBlocks"] = []any{}
	}

	if value != nil {
		if eval, ok := value.(interface{ Eval(any) (any, error) }); ok {
			evaluated, err := eval.Eval(context)
			if err != nil {
				return nil, err
			}
			value = evaluated
		}
	}

	if len(rules) > 0 {
		// Assuming that there is only one rule at this point - that is how parser constructs the rule
		if eval, ok := rules[0].(interface{ Eval(any) (*Ruleset, error) }); ok {
			evaluated, err := eval.Eval(context)
			if err != nil {
				return nil, err
			}
			rules = []any{evaluated}
			evaluated.Root = true
		}
	}

	// Restore media bubbling information
	if ctx, ok := context.(map[string]any); ok {
		ctx["mediaPath"] = mediaPathBackup
		ctx["mediaBlocks"] = mediaBlocksBackup
	}

	return NewAtRule(a.Name, value, rules, a.GetIndex(), a.FileInfo(), a.DebugInfo, a.IsRooted, a.VisibilityInfo()), nil
}

// Variable returns a variable from the first rule (if rules exist)
func (a *AtRule) Variable(name string) any {
	if len(a.Rules) > 0 {
		// Assuming that there is only one rule at this point - that is how parser constructs the rule
		if ruleset, ok := a.Rules[0].(*Ruleset); ok {
			return ruleset.Variable(name)
		}
	}
	return nil
}

// Find finds rules matching a selector (delegates to first rule if exists)
func (a *AtRule) Find(selector any, self any, filter func(any) bool) []any {
	if len(a.Rules) > 0 {
		// Assuming that there is only one rule at this point - that is how parser constructs the rule
		if ruleset, ok := a.Rules[0].(*Ruleset); ok {
			return ruleset.Find(selector, self, filter)
		}
	}
	return nil
}

// Rulesets returns rulesets from the first rule (if rules exist)
func (a *AtRule) Rulesets() []any {
	if len(a.Rules) > 0 {
		// Assuming that there is only one rule at this point - that is how parser constructs the rule
		if ruleset, ok := a.Rules[0].(*Ruleset); ok {
			return ruleset.Rulesets()
		}
	}
	return nil
}

// OutputRuleset outputs CSS for rules with proper formatting
func (a *AtRule) OutputRuleset(context any, output *CSSOutput, rules []any) {
	ruleCnt := len(rules)
	
	ctx, ok := context.(map[string]any)
	if !ok {
		ctx = make(map[string]any)
	}
	
	tabLevel := 0
	if tl, ok := ctx["tabLevel"].(int); ok {
		tabLevel = tl
	}
	tabLevel = tabLevel + 1
	ctx["tabLevel"] = tabLevel

	compress := false
	if c, ok := ctx["compress"].(bool); ok {
		compress = c
	}

	if compress {
		output.Add("{", nil, nil)
		for i := 0; i < ruleCnt; i++ {
			if gen, ok := rules[i].(interface{ GenCSS(any, *CSSOutput) }); ok {
				gen.GenCSS(context, output)
			}
		}
		output.Add("}", nil, nil)
		ctx["tabLevel"] = tabLevel - 1
		return
	}

	// Non-compressed
	// JavaScript: Array(context.tabLevel).join('  ') creates (tabLevel-1) pairs of spaces
	tabSetStr := "\n" + strings.Repeat("  ", tabLevel-1)
	tabRuleStr := tabSetStr + "  "

	if ruleCnt == 0 {
		output.Add(" {"+tabSetStr+"}", nil, nil)
	} else {
		output.Add(" {"+tabRuleStr, nil, nil)

		// Process rules, setting lastRule for the final rule
		for i := 0; i < ruleCnt; i++ {
			if i > 0 {
				output.Add(tabRuleStr, nil, nil)
			}

			// Set lastRule flag for the last rule (similar to JavaScript ruleset.js line 533)
			if i+1 == ruleCnt {
				ctx["lastRule"] = true
			}

			if gen, ok := rules[i].(interface{ GenCSS(any, *CSSOutput) }); ok {
				gen.GenCSS(context, output)
			}

			// Clear lastRule after processing
			if i+1 == ruleCnt {
				ctx["lastRule"] = false
			}
		}

		output.Add(tabSetStr+"}", nil, nil)
		// Add newline after closing brace to separate from next top-level rule
		output.Add("\n", nil, nil)
	}

	ctx["tabLevel"] = tabLevel - 1
}

// SetAllExtends sets the AllExtends field (used by ExtendFinderVisitor)
func (a *AtRule) SetAllExtends(extends []*Extend) {
	a.AllExtends = extends
}

// GetAllExtends returns the AllExtends field (used by ProcessExtendsVisitor)
func (a *AtRule) GetAllExtends() []*Extend {
	return a.AllExtends
} 